dpnp/dpnp_iface_logic.py

```diff
@@ -572,48 +546,109 @@ def any(a, /, axis=None, out=None, keepdims=False, *, where=True):
+        Second input array, also expected to have numeric data type.
+        Both inputs `a` and `b` can not be scalars at the same time.
+    rtol : {dpnp.ndarray, usm_ndarray, scalar}, optional
+        The relative tolerance parameter. Default: ``1e-05``.
+    atol : {dpnp.ndarray, usm_ndarray, scalar}, optional
+        The absolute tolerance parameter. Default: ``1e-08``.
+    equal_nan : bool
+        Whether to compare ``NaNs`` as equal. If ``True``, ``NaNs`` in `a` will
+        be considered equal to ``NaNs`` in `b` in the output array.
+
+    Returns
+    -------
+    out : dpnp.ndarray
+        Returns a boolean array of where `a` and `b` are equal within the given
+        tolerance.
 
     See Also
     --------
-    :obj:`dpnp.allclose` : Returns True if two arrays are element-wise equal
-                           within a tolerance.
+    :obj:`dpnp.allclose` : Returns ``True`` if two arrays are element-wise
+                           equal within a tolerance.
 
     Examples
     --------
     >>> import dpnp as np
-    >>> x1 = np.array([1e10,1e-7])
-    >>> x2 = np.array([1.00001e10,1e-8])
-    >>> out = np.isclose(x1, x2)
-    >>> [i for i in out]
-    [True, False]
+    >>> a = np.array([1e10, 1e-7])
+    >>> b = np.array([1.00001e10, 1e-8)
+    >>> np.isclose(a, b)
+    array([ True, False])
 
-    """
+    >>> a = np.array([1e10, 1e-8])
+    >>> b = np.array([1.00001e10, 1e-9])
+    >>> np.isclose(a, b)
+    array([ True,  True])
 
-    # x1_desc = dpnp.get_dpnp_descriptor(x1)
-    # x2_desc = dpnp.get_dpnp_descriptor(x2)
-    # if x1_desc and x2_desc:
-    #     result_obj = dpnp_isclose(
-    #         x1_desc, x2_desc, rtol, atol, equal_nan
-    #     ).get_pyobj()
-    #     return result_obj
+    >>> a = np.array([1e10, 1e-8])
+    >>> b = np.array([1.0001e10, 1e-9])
+    >>> np.isclose(a, b)
+    array([False,  True])
 
-    return call_origin(
-        numpy.isclose, x1, x2, rtol=rtol, atol=atol, equal_nan=equal_nan
+    >>> a = np.array([1.0, np.nan])
+    >>> b = np.array([1.0, np.nan])
+    >>> np.isclose(a, b)
+    array([ True, False])
+    >>> np.isclose(a, b, equal_nan=True)
+    array([ True,  True])
+
+    >>> a = np.array([0.0, 0.0])
+    >>> b = np.array([1e-8, 1e-7])
+    >>> np.isclose(a, b)
+    array([True, False])
+    >>> b = np.array([1e-100, 1e-7])
+    >>> np.isclose(a, b, atol=0.0)
+    array([False, False])
+
+    >>> a = np.array([1e-10, 1e-10])
+    >>> b = np.array([1e-20, 0.0])
+    >>> np.isclose(a, b)
+    array([ True,  True])
+    >>> b = np.array([1e-20, 0.999999e-10])
+    >>> np.isclose(a, b atol=0.0)
+    array([False,  True])
+
+    """
+
+    dpnp.check_supported_arrays_type(a, b, scalar_type=True)
+    dpnp.check_supported_arrays_type(
+        rtol, atol, scalar_type=True, all_scalars=True
     )
 
+    if dpnp.isscalar(b):
+        dt = dpnp.result_type(
+            b, dpnp.default_float_type(sycl_queue=a.sycl_queue)
+        )
+        b = dpnp.asarray(
```

Comment:
That might cause an issue on Iris Xe:
```python
import dpnp

a = dpnp.array([1])
dpnp.isclose(a, 10**5)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[4], line 1
----> 1 dpnp.isclose(a, 10**5)

File ~/code/dpnp/dpnp/dpnp_iface_logic.py:635, in isclose(a, b, rtol, atol, equal_nan)
    631 if dpnp.isscalar(b):
    632     dt = dpnp.result_type(
    633         b, dpnp.default_float_type(sycl_queue=a.sycl_queue)
    634     )
--> 635     b = dpnp.asarray(
    636         b, dtype=dt, sycl_queue=a.sycl_queue, usm_type=a.usm_type
    637     )
    638 elif b.dtype.kind == "i":
    639     dt = dpnp.result_type(
    640         b, dpnp.default_float_type(sycl_queue=b.sycl_queue)
    641     )

File ~/code/dpnp/dpnp/dpnp_iface_arraycreation.py:517, in asarray(a, dtype, order, like, device, usm_type, sycl_queue)
    439 """
    440 Converts an input object into array.
    441
   (...)
    512
    513 """
    515 dpnp.check_limitations(like=like)
--> 517 return dpnp_container.asarray(
    518     a,
    519     dtype=dtype,
    520     order=order,
    521     device=device,
    522     usm_type=usm_type,
    523     sycl_queue=sycl_queue,
    524 )

File ~/code/dpnp/dpnp/dpnp_container.py:122, in asarray(x1, dtype, copy, order, device, usm_type, sycl_queue)
    116     x1_obj = x1
    118 sycl_queue_normalized = dpnp.get_normalized_queue_device(
    119     x1_obj, device=device, sycl_queue=sycl_queue
    120 )
--> 122 array_obj = dpt.asarray(
    123     x1_obj,
    124     dtype=dtype,
    125     copy=copy,
    126     order=order,
    127     usm_type=usm_type,
    128     sycl_queue=sycl_queue_normalized,
    129 )
    131 # return x1 if dpctl returns a zero copy of x1_obj
    132 if array_obj is x1_obj and isinstance(x1, dpnp_array):

File ~/miniforge3/envs/dpnp_dev/lib/python3.11/site-packages/dpctl/tensor/_ctors.py:648, in asarray(obj, dtype, device, copy, usm_type, sycl_queue, order)
    644     raise ValueError(
    645         f"Converting {type(obj)} to usm_ndarray requires a copy"
    646     )
    647 # obj is a scalar, create 0d array
--> 648 return _asarray_from_numpy_ndarray(
    649     np.asarray(obj, dtype=dtype),
    650     dtype=dtype,
    651     usm_type=usm_type,
    652     sycl_queue=sycl_queue,
    653     order="C",
    654 )

File ~/miniforge3/envs/dpnp_dev/lib/python3.11/site-packages/dpctl/tensor/_ctors.py:253, in _asarray_from_numpy_ndarray(ary, dtype, usm_type, sycl_queue, order)
    249     res = dpt.usm_ndarray(
    250         res.shape, dtype=res.dtype, buffer=res.usm_data, strides=new_strides
    251     )
    252 else:
--> 253     _ensure_native_dtype_device_support(dtype, copy_q.sycl_device)
    254     res = dpt.usm_ndarray(
    255         ary.shape,
    256         dtype=dtype,
   (...)
    259         buffer_ctor_kwargs={"queue": copy_q},
    260     )
    261 res[...] = ary

File ~/miniforge3/envs/dpnp_dev/lib/python3.11/site-packages/dpctl/tensor/_ctors.py:280, in _ensure_native_dtype_device_support(dtype, dev)
    266 """Check that dtype is natively supported by device.
    267
    268 Arg:
   (...)
    277         if device does not natively support this `dtype`.
    278 """
    279 if dtype in [dpt.float64, dpt.complex128] and not dev.has_aspect_fp64:
--> 280     raise ValueError(
    281         f"Device {dev.name} does not provide native support "
    282         "for double-precision floating point type."
    283     )
    284 if (
    285     dtype
    286     in [
   (...)
    289     and not dev.has_aspect_fp16
    290 ):
    291     raise ValueError(
    292         f"Device {dev.name} does not provide native support "
    293         "for half-precision floating point type."
    294     )

ValueError: Device Intel(R) Graphics [0x9a49] does not provide native support for double-precision floating point type.
```
I would propose to calculate dt here based on `a` also:
```python
        dt = dpnp.result_type(
            a, b, dpnp.default_float_type(sycl_queue=a.sycl_queue)
        )
```